utils/src/array.rs

```diff
@@ -7,33 +7,36 @@
 //! (non-fixed-sized) vectors.
 
 /// Converts a two-dimensional vector to a constant sized two-dimensional array.
-pub fn vec_to_boxed_array2<T, const N: usize, const M: usize>(
+pub fn vec_to_boxed_array2<T: Clone, const N: usize, const M: usize>(
     vec: Vec<Vec<T>>,
 ) -> Box<[[T; N]; M]> {
-    let vec_of_slices2: Vec<[T; N]> = vec
-        .into_iter()
-        .map(|x: Vec<T>| {
-            let y: Box<[T]> = x.into_boxed_slice();
-            let z: Box<[T; N]> = y
-                .try_into()
-                .unwrap_or_else(|_| panic!("vec_to_boxed_array2: length mismatch inner array"));
-            *z
-        })
-        .collect();
-    let array: Box<[[T; N]; M]> = vec_of_slices2
+    let mut vec_of_slices2: Vec<[T; N]> = vec![];
+    vec.into_iter().for_each(|x: Vec<T>| {
+        let y: Box<[T]> = x.into_boxed_slice();
+        let z: Box<[T; N]> = y
+            .try_into()
+            .unwrap_or_else(|_| panic!("vec_to_boxed_array2: length mismatch inner array"));
+        let zz: &[[T; N]] = std::slice::from_ref(z.as_ref());
+        vec_of_slices2.extend_from_slice(zz);
```

Comment:
From https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extend_from_slice
```
Iterates over the slice other, clones each element, and then appends it to this Vec. The other slice is traversed in-order.
```
Then the `Clone` requirement.

Reply:
Yep, exactly. Although I think "clones each element" here is misleading: the description seems to imply that it does put the element on stack, but it actually extends directly from memory. Which was the only way I found to add anything to a vector directly from memory.